api/src/main/java/ai/djl/ndarray/gc/WeakHashMapWrapper.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.ndarray.gc;
+
+import ai.djl.ndarray.NDArray;
+
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.reflect.Proxy;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+/**
+ * {@code WeakHashMapWrapper<K, V>} wraps a {@link WeakHashMap}. It has a {@link ReferenceQueue} to
+ * get informed by the garbage collector. On method invocations it looks for messages from the
+ * garbage collector and removes the corresponding entries.
+ */
+public class WeakHashMapWrapper<K, V> implements Map<K, V> {
+
+    private final WeakHashMap<K, V> map = new WeakHashMap<>();
+    private final ReferenceQueue<Object> queue = new ReferenceQueue<>();
+
+    private final Set<WeakReferenceWrapper<K, V>> weakReferenceWrapperSet = new HashSet<>();
+
+    private void checkQueue() {
+        for (Reference<?> ref; (ref = queue.poll()) != null; ) {
+            synchronized (queue) {
```

Comment:
This implementation may have big performance impact:
1. every NDArray api call (DynamicInvocationHandler.invoke), will call checkQueue() method
2. in this call, it loop through the reference queue
3. this is synchronized call, will cause thread contention in heavy multithreading case

Reply:
I completely agree that checkQueue() must not become a performance bottleneck. 
Fortunately the ReferenceQueue is made such that there is no "looping through the reference queue" on each call. Let's look together at the ReferenceQueue source code: 

![grafik](https://user-images.githubusercontent.com/9622808/211319497-b79cda40-20f3-4711-b2d2-23e2004d9070.png)

If there are no objects on the queue it immediately returns. If there are objects all objects are removed on checkQueue() such that the next time (without garbage collection inbetween) it again returns immediately.

If for whatever reason the performance becomes worse and not much better as expected by switching on gc, the user may decide not to use this opt-in feature.